cmd/client/client.go

```diff
@@ -0,0 +1,216 @@
+		prefix = strings.Join([]string{"%{color}", prefix, "%{color:reset}"}, "")
+	}
+	logging.SetFormatter(logging.MustStringFormatter(strings.Join([]string{prefix, " %{message}"}, "")))
+	logging.SetBackend(logging.NewLogBackend(os.Stderr, "", 0))
+	logger = logging.MustGetLogger("")
+}
+
+type serverConfig struct {
+	host   string
+	port   int
+	cipher string
+	secret string
+}
+
+func parseKey(k string) (serverConfig, error) {
+	u, err := url.Parse(k)
+	if err != nil {
+		return serverConfig{}, err
+	}
+
+	port, err := strconv.Atoi(u.Port())
+	if err != nil {
+		return serverConfig{}, fmt.Errorf("invalid port: %v", err)
+	}
+
+	secret := u.User.String()
+	if !strings.Contains(secret, ":") {
+		b, err := base64.StdEncoding.DecodeString(secret)
+		if err != nil {
+			b, err = base64.RawStdEncoding.DecodeString(u.User.String())
+			if err != nil {
+				return serverConfig{}, fmt.Errorf("invalid password in key: %v", err)
+			}
+		}
+		secret = string(bytes.TrimSpace(b))
+	}
+	p := strings.Split(secret, ":")
+	if len(p) != 2 {
+		return serverConfig{}, fmt.Errorf("invalid password in key")
+	}
+
+	return serverConfig{
+		host:   u.Hostname(),
+		port:   port,
+		cipher: p[0],
+		secret: p[1],
+	}, nil
+}
+
+func resolveHostPort(addr string) (string, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return "", err
+	}
+	ip, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return "", fmt.Errorf("resolving hostname failed: %v", err)
+	}
+	return net.JoinHostPort(ip.String(), port), err
+}
+
+type SocksSSClient struct {
+	config   serverConfig
+	listener *net.TCPListener
+}
+
+// RunSocksSSClient starts a SOCKS server which proxies connections to the specified shadowsocks server.
+func RunSocksSSClient(bindAddr string, listenPort int, config serverConfig) (*SocksSSClient, error) {
+	listener, err := net.ListenTCP("tcp", &net.TCPAddr{IP: net.ParseIP(bindAddr), Port: listenPort})
+	if err != nil {
+		return nil, fmt.Errorf("listenTCP failed: %v", err)
+	}
+	logger.Infof("Listenting at %v", listener.Addr())
+
+	d, err := client.NewClient(config.host, config.port, config.secret, config.cipher)
+	if err != nil {
+		return nil, fmt.Errorf("failed connecting to server: %v", err)
+	}
+
+	go func() {
+		for {
+			clientConn, err := listener.AcceptTCP()
+			if err != nil {
+				if errors.Is(err, net.ErrClosed) {
+					logger.Info("SOCKS listener closed")
+				} else {
+					logger.Errorf("Accepting SOCKS connection failed: %v\n", err)
+				}
+				break
+			}
+			go func() {
+				defer clientConn.Close()
+
+				tgtAddr, err := socks.Handshake(clientConn)
+				if err != nil {
+					logger.Errorf("SOCKS handshake failed: %v", err)
+					return
+				}
+				addr, err := resolveHostPort(tgtAddr.String())
+				if err != nil {
+					logger.Errorf("Failed to resolve target address: %v", err)
+					return
+				}
+
+				logger.Debugf("Opening connection for %s", addr)
+				targetConn, err := d.DialTCP(nil, addr)
```

Comment:
We need to make it clear what address we are talking about.
```suggestion
				targetConn, err := d.DialTCP(nil, tgtAddr)
```